### PR TITLE
FE-6858 - update date-fns from ^2.30.0 to ^4.1.0

### PR DESCRIPTION
**Pull request by quinnturner**
Closes: https://github.com/Sage/carbon/issues/7004

### Proposed behaviour

Updates date-fns

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Playwright automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Related docs have been updated if required

#### QA

- [ ] Tested in provided StackBlitz sandbox/Storybook
- [ ] Add new Playwright test coverage if required
- [x] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Additional context

The main concern with my current implementation is that `<DayPicker` has a property locale that accepts a `date-fns` localize type.

`weekdaysLong` and `weekdaysShort` are now not accepted properties.

I did not know what to do in this scenario, so I removed the code entirely. However, I suspect that's not the correct behaviour.

I would appreciate guidance on how to proceed.

### CODE REVIEW

**Review by damienrobson-sage — APPROVED** [review]

File: src/components/date/__internal__/date-picker/date-picker.component.tsx [hunk 1]
```diff
@@ -245,9 +238,6 @@ export const DatePicker = ({
             locale={{
               localize: {
                 ...defaultLocale.localize,
-                months: monthsLong,
```
Comment (quinnturner) [inline src/components/date/__internal__/date-picker/date-picker.component.tsx:248]: I really am not sure what to do here. There's a good chance this isn't the best behaviour. Looking for advice/opinions.

Comment (Parsium) [reply src/components/date/__internal__/date-picker/date-picker.component.tsx:248]: The visible labels and ARIA labels look consistent with production, so this should be fine. We have passing tests that check the weekday labels are two characters long in German and Polish, so this edge case is covered :)

**Review by Parsium — APPROVED** [review]
Thanks for contributing @quinnturner! This should eliminate another piece of tech debt for us. I'll mark this as ready for QA.

Before we release, since this involves a major version upgrade of a dependency, we should mark it as a breaking change. Could you update the commit message to reflect this please? The message will be used in our release notes.

Here's a suggestion:

```suggestion
chore: upgrade `date-fns` to version ^4.1.0


BREAKING CHANGE: `date-fns` is now ESM-first. While CommonJS
remains supported, this change may cause issues in some environments.
If you encounter issues, you may need to update your bundler
configuration to support ESM code.
```

**Comment by quinnturner** [comment]
Thanks, @Parsium; I've amended the commit message to match yours verbatim.

**Comment by edleeks87** [comment]
Do not merge until https://github.com/Sage/carbon/pull/7148 has been released
